crates/cli-support/src/js/mod.rs

```diff
@@ -490,6 +490,45 @@ impl<'a> Context<'a> {
         let mut dst = format!("class {} {{\n", name);
         let mut ts_dst = format!("export {}", dst);
 
+        let (mkweakref, freeref) = if self.config.weak_refs {
+            // When weak refs are enabled we use them to automatically free the
+            // contents of an exported rust class when it's gc'd. Note that a
+            // manual `free` function still exists for deterministic
+            // destruction.
+            //
+            // This is implemented by using a `WeakRefGroup`-per-exported-class
```

Comment:
Why do you create a group per class? To support different protocols for freeing instances? It doesn't look like there are differences in protocol right now, so ideally we'd just have one group per WebAssembly instance.

The idea with `WeakRefGroup` is to use it to guarantee cleanup as long as an interconnected system hasn't been GC'd. Specifically, for a WebAssembly module instance, you'd tie the group's lifetime to the instance's lifetime—most likely by just setting a property on the instance pointing to the group. If the entire instance is GC'd, then there's no need to do individual cleanup, so the group should be GC'd without running finalizers, too. Conversely, as long as the group is around, finalizers are guaranteed to run: the group keeps all the `WeakRef` instances created through it alive.

Reply:
> Why do you create a group per class? To support different protocols for freeing instances? It doesn't look like there are differences in protocol right now, so ideally we'd just have one group per WebAssembly instance.

They have different destructors, and therefore each class has different drop-and-free functions. See my comment elsewhere in the PR about how we could have a single group, though.

Reply:
Ah yeah I wasn't initially sure on the expense of having a `WeakRefGroup`, but using just one is totally plausible!